### PR TITLE
Minecraft 1.20.2 support

### PR DESCRIPTION
**Pull request by rom1504**
migrated from https://github.com/PrismarineJS/mineflayer/pull/3243

### CODE REVIEW

File: lib/plugins/entities.js [hunk 1]
```diff
@@ -212,6 +213,14 @@ function inject (bot) {
     entity.yaw = conv.fromNotchianYawByte(packet.yaw)
     entity.pitch = conv.fromNotchianPitchByte(packet.pitch)
     entity.objectData = packet.objectData
+
+    if (entity.type === 'player') {
```
Comment (rom1504) [inline lib/plugins/entities.js:217]: that looks pretty suspicious, not sure if it's right

Comment (wgaylord) [reply lib/plugins/entities.js:217]: I suspect it is correct based on existing code using the opposite condition. https://github.com/PrismarineJS/mineflayer/blob/master/lib/plugins/entities.js#L42

Not sure if what its being used for is correct tho.

**Comment by wgaylord** [comment]
I think the only thing left is to fix what ever is breaking the External Bed and Elytra tests. Then fix the Internal tests since they are all totally failing.

**Comment by wgaylord** [comment]
Found https://github.com/PrismarineJS/minecraft-data/issues/817 to be why both the External Bed and Elytra tests are failing

File: lib/plugins/entities.js (hunk 1 above)
Comment (extremeheat) [reply lib/plugins/entities.js:216]: named_entity_spawn was removed, so players spawn with the generic spawn_entity packet. However there are alot of unhandled fields that are not currently accounted for - https://github.com/PrismarineJS/mineflayer/blob/a0a577763d08f44494669793f8f4a2f8b2d58e0f/lib/plugins/entities.js#L132

**Comment by wgaylord** [comment]
okay, so the major thing stopping this is changes to the entity spawning packet in both Mineflayer and in the internal tests
https://github.com/PrismarineJS/mineflayer/blob/mc-1.20.2/lib/plugins/entities.js#L132-L157 is how we used to handle player spawns. For 1.20.2+ players spawn using the same as other mobs, partially implemented here https://github.com/PrismarineJS/mineflayer/blob/mc-1.20.2/lib/plugins/entities.js#L196-L224 We need to fill out the rest of the fields for the player.

This same sort of change needs to happen on in the integral tests based on version.  Where https://github.com/PrismarineJS/mineflayer/blob/mc-1.20.2/test/internalTest.js#L635-L647 and https://github.com/PrismarineJS/mineflayer/blob/mc-1.20.2/test/internalTest.js#L700-L712 need to be setup so if we are not 1.20.2+ they are used and if we are 1.20.2+ something like https://github.com/PrismarineJS/mineflayer/blob/mc-1.20.2/test/internalTest.js#L786-L800 is used (of course with the correct values.)

**Comment by rom1504** [comment]
@wgaylord anything unclear about it or you just need to do it?

**Comment by wgaylord** [comment]
Mostly just need to do it, wanted to document exactly what was still needed, incase I don't get the time to work on it.

**Comment by extremeheat** [comment]
Ok, should be passing now. Required changes were:

* server.on `login` (packet) -> `playerJoin` (nmp event) to handle play state
* named_entity_spawn is now done with spawn_entity packet
* dimension codec is sent outside login packet now (before play state)

File: test/internalTest.js [hunk 2]
```diff
@@ -632,17 +632,34 @@ for (const supportedVersion of mineflayer.testedVersions) {
...
-            playerUUID: '1-2-3-4',
-            x: 1,
-            y: 2,
-            z: 3,
-            yaw: 0,
-            pitch: 0,
-            currentItem: -1,
-            metadata: []
-          })
+          if (bot.registry.version['>=']('1.20.2')) {
```
Comment (rom1504) [inline test/internalTest.js:635]: would be good to make that a feature

**Comment by rom1504** [comment]
great

File: test/internalTest.js (hunk 2 above)
Comment (extremeheat) [reply test/internalTest.js]: done, looks like removing dupe https://github.com/PrismarineJS/minecraft-data/commit/587a1341c83ef4fc1cf3353d35b09435a11f8217 caused range to be incorrect, just fixed that

**Comment by rom1504** [comment]
Remaining here, test some examples manually
* [ ] Inventory.js
* [ ] digger
* [ ] chest
* [ ] check previous PRs

**Comment by rom1504** [comment]
I tested digger, chest, inventory, jumper, viewer and pathfinder, all working

**Pull request merged**